### PR TITLE
Deprecate and move to Internal factories methods

### PR DESCRIPTION
**Pull request by aledustet**
Why:
These methods are used internally for the functionality of the library
and are subject to change. Therefore shouldn't be part of the public
interface.

This PR:
- Moves the ```register_factory``` and ```factory_by_name``` methods to
the ```FactoryBot::Internal``` class.
- Deprecates the use of ```register_factory``` and ```factory_by_name```
from the ```FactoryBot``` module.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,3 +1,4 @@
+*.DS_Store
```
Comment (composerinteralia) [inline .gitignore]: I have always felt that OS and editor specific belong in each person's own global gitignore (checkout the [thoughtbot dotfiles global gitignore](https://github.com/thoughtbot/dotfiles/blob/master/gitignore), for example) rather than in the project's gitignore.

That said, this file already has `*.swp` in it, which is a vim thing.

Totally your call on whether you want to include it.

so I guess I don't feel that strongly one way or another

File: lib/factory_bot.rb [hunk 2]
```diff
@@ -75,19 +75,21 @@ def self.lint(*args)
   end
 
   class << self
-    delegate :factories,
+    delegate :callbacks,
```
Comment (composerinteralia) [inline lib/factory_bot.rb]: 👍 🔤

File: lib/factory_bot.rb [hunk 3]
```diff
@@ -98,6 +100,8 @@ class << self
 
     deprecate :allow_class_lookup,
```
Comment (composerinteralia) [inline lib/factory_bot.rb]: Are we also able to deprecate `FactoryBot.factories`?

File: lib/factory_bot/null_object.rb [hunk 4]
```diff
@@ -6,11 +6,8 @@ def initialize(methods_to_respond_to)
     end
 
     def method_missing(name, *args, &block)
-      if respond_to?(name)
```
Comment (composerinteralia) [inline lib/factory_bot/null_object.rb]: This seems unrelated to this PR. Also, I think I like the first version a little better. It is easier to see the conditional.

File: spec/factory_bot/internal_spec.rb [hunk 5]
```diff
@@ -8,6 +8,11 @@
         from(0).
         to(1)
     end
+
+    it "returns the registered trait" do
```
Comment (composerinteralia) [inline spec/factory_bot/internal_spec.rb]: Is it important that it returns the registered trait? Would it matter if we eventually return something else? If not, we might not need this test.

File: spec/factory_bot/internal_spec.rb [hunk 6]
```diff
@@ -27,6 +32,11 @@
         from(0).
         to(1)
     end
+
+    it "returns the registered sequence" do
```
Comment (composerinteralia) [inline spec/factory_bot/internal_spec.rb]: Same question here

File: spec/factory_bot/internal_spec.rb [hunk 7]
```diff
@@ -48,4 +58,28 @@
       FactoryBot::Internal.rewind_sequences
     end
   end
+
+  describe ".register_factory" do
```
Comment (composerinteralia) [inline spec/factory_bot/internal_spec.rb]: As mentioned in https://github.com/thoughtbot/factory_bot/pull/1291#discussion_r283066959, our style guide (and also I 😄 ) would prefer some newlines in these tests.

File: spec/factory_bot/internal_spec.rb [hunk 8]
```diff
@@ -48,4 +58,28 @@
...
+    it "registers the provided factory" do
+      factory = FactoryBot::Factory.new(:object)
+      configuration = FactoryBot::Internal.configuration
+      expect { FactoryBot::Internal.register_factory(factory) }.
+        to change { configuration.factories.count }.
+        from(0).
+        to(1)
+    end
+
+    it "returns the registered factory" do
```
Comment (composerinteralia) [inline spec/factory_bot/internal_spec.rb]: And here

File: .gitignore (hunk 1 above)
Comment (aledustet) [reply .gitignore]: I was wondering why it wasn't there. I have it in my global config but I was editing it home and it slipped I agree that specific settings should belong in your own environment. I'll remove it 👍

File: spec/factory_bot/internal_spec.rb (hunk 5 above)
Comment (aledustet) [reply spec/factory_bot/internal_spec.rb]: ```register_trait``` and ```register_sequence``` both have the same signature in the method, I thought it was necessary, that's why I added the test, I only see it used in ```FactoryBot::Syntax::Default::DSL``` and the result of using the block in the DSL is not documented(to the best of my knowledge) to use the return value. Maybe we are safe removing the return statement? There is a test under ```attributes/dynamic_spec``` that uses the return value.

File: lib/factory_bot.rb (hunk 3 above)
Comment (aledustet) [reply lib/factory_bot.rb]: There is one mention of using this method in the [docs](https://github.com/thoughtbot/factory_bot/blob/master/GETTING_STARTED.md#linting-factories), maybe the use case it's not needed?
